### PR TITLE
Parallel negf

### PR DESCRIPTION
**Pull request by pecchia**
Transport extensions to be merged. 
I'm wearing my rock helmet...

### CODE REVIEW

File: doc/dftb+/manual/transport.tex [hunk 1]
```diff
@@ -0,0 +1,1007 @@
...
+\item[\is{FirstLayerAtoms}] defines the first atom of PLs in the device
+  region. By default there is only one layer (the entire device
+  region). Alternatively the user can manually reorder and partition the
+  structure into layers for efficient GF calculations.
+  
+  Differently from the contact PLs, the device layers do not need to represent
+  unit cell repetitions. The device geometry must be manually ordered in such a
+  way that all the atoms within each layer are contiguous and adjacent layers
+  are placed next to each other. This ensures that the constructed Hamiltonian
+  and Overlap are block tri-diagonal. Refer to \cite{Pecchia_NJP} for a
```
Comment (bhourahine) [inline doc/dftb+/manual/transport.tex]: Undefined citation, missing from bibliography

File: doc/dftb+/manual/transport.tex [hunk 2]
```diff
@@ -0,0 +1,1007 @@
...
+  electrostatics must be calculated (i.e. the \is{Electrostatics = Poisson}
+  should not be declared).
+
+
+\section{Contour integration}
+
+Much of the computational work is in the integration of the energy resolved
+density matrix, represented via the NEGF matrix. The integration is efficiently
+performed with a complex contour integration and a real axis integration, as
+shown in Figure~\ref{fig:contour} and discussed in references \cite{Pecchia_RPP,
```
Comment (bhourahine) [inline doc/dftb+/manual/transport.tex]: Missing reference

File: prog/dftb+/prg_dftb/main.F90 [hunk 3]
```diff
@@ -236,29 +258,64 @@ subroutine runDftbPlus(env)
...
       call setUpExternalElectricField(tEField, tTDEField, tPeriodic, EFieldStrength,&
           & EFieldVector, EFieldOmega, EFieldPhase, neighborList, nNeighbor, iCellVec,&
           & img2CentCell, cellVec, deltaT, iGeoStep, coord0Fold, coord, EField,&
           & potential%extAtom(:,1), absEField)
-    
+
+
       call mergeExternalPotentials(orb, species, potential)
+        
+      ! Dirty trick to jump out in case of non-scc calculations with transport only
```
Comment (bhourahine) [inline prog/dftb+/prg_dftb/main.F90]: Can't this set maxSccIter in the same way the regular non-SCC calculations deal with this?

File: prog/dftb+/prg_dftb/main.F90 [hunk 4]
```diff
@@ -915,6 +1045,30 @@ subroutine handleCoordinateChange(env, coord0, latVec, invLatVec, species0, mCut
...
+    integer, intent(in) :: iAtomStart(:)
+    type(TNeighborList), intent(in) :: neighborList
+    integer, intent(in) :: nNeighbor(:)
+ 
+    ! known issue about the PLs: We need an authomatic partitioning
+    call negf_init_csr(iAtomStart, neighborList%iNeighbor, nNeighbor, img2CentCell, orb)
+        
+    call negf_init_str(negfStr, transpar, ginfo%greendens, neighborList%iNeighbor, nNeighbor, img2CentCell)
+        
+    call negf_init_dephasing(ginfo%tundos)  !? why tundos
```
Comment (bhourahine) [inline prog/dftb+/prg_dftb/main.F90]: Assume this comment is asking if there is any reason not to pass in the whole structure?

File: prog/dftb+/lib_poisson/gconstants.F90 [hunk 5]
```diff
@@ -0,0 +1,15 @@
+module gconstants
```
Comment (bhourahine) [inline prog/dftb+/lib_poisson/gconstants.F90]: All of these are either redundant, or should be in the constants.F90

File: prog/dftb+/lib_extlibs/mat_conv.F90 [hunk 6]
```diff
@@ -0,0 +1,810 @@
...
+  use Constants, only : pi
+  use CommonTypes, only : TOrbitals
+  implicit none
+  private
+
+  public :: init,   destruct
+  public :: foldToCSR, unfoldFromCSR
+  public :: TDftbMat_r, TDftbMat_z
+
+  type TDftbMat_r
```
Comment (bhourahine) [inline prog/dftb+/lib_extlibs/mat_conv.F90]: Most of these pointers probably can be converted to allocatables

File: prog/dftb+/lib_extlibs/negf_int.F90 [hunk 7]
```diff
@@ -0,0 +1,2117 @@
...
+  public :: calcdensity_green
+  public :: calcEdensity_green
+  public :: calcPDOS_green
+  public :: calc_current
+  public :: local_currents
+  
+  ! interface csr matrices. The pattering must be predefined 
+  ! using negf_init_csr 
+  public :: negf_init_csr
+  type(z_CSR) :: csrHam, csrOver
```
Comment (bhourahine) [inline prog/dftb+/lib_extlibs/negf_int.F90]: Needs to become derived type variables

File: prog/dftb+/lib_extlibs/negf_int.F90 [hunk 8]
```diff
@@ -0,0 +1,2117 @@
...
+    ! It affects especially transmission that is not accurate more than 
+    ! this value.
+    call set_drop(1.d-20)
+
+ 
+    ! ------------------------------------------------------------------------------
+    ! Assign spin degenracy and check consistency between different input blocks
+    if (tundos%defined .and. greendens%defined) then
+      if (tundos%gSpin .ne. greendens%gSpin) then
+        write(*,*) "ERROR: spin degeneracy is not consistent between different input blocks"
```
Comment (bhourahine) [inline prog/dftb+/lib_extlibs/negf_int.F90]: Write(*, statements won't connect with the parallel write-out structure of the main code - things like should this be appearing on rank=0 or from anything in the comm world?

File: prog/dftb+/lib_extlibs/negf_int.F90 [hunk 9]
```diff
@@ -0,0 +1,2117 @@
...
+    else
+      ncont = 0
+    endif
+    
+    ! ------------------------------------------------------------------------------
+    !! Set defaults and fill up the parameter structure with them
+    call init_negf(negf)
+    call init_contacts(negf, ncont)
+    call set_scratch(negf, ".")
+    if (id0) call create_scratch(negf)
```
Comment (bhourahine) [inline prog/dftb+/lib_extlibs/negf_int.F90]: id0 is coming back from libNEGF, probably should be using internal DFTB+ variables (there is a logical in the environment module for this).

File: prog/dftb+/prg_dftb/main.F90 [hunk 10]
```diff
@@ -1316,8 +1521,22 @@ subroutine addChargePotentials(env, sccCalc, qInput, q0, chargePerShell, orb, sp
...
+    !> whether contacts are uploaded
+    logical, intent(in) :: tUpload
+
+    !> uploded potential per shell per atom
+    real(dp), intent(in) :: shiftPerLUp(:,:)
+
+    ! local variables
     real(dp), allocatable :: atomPot(:,:)
     real(dp), allocatable :: shellPot(:,:,:)
+    real(dp), allocatable, save :: shellPotBk(:,:)
```
Comment (bhourahine) [inline prog/dftb+/prg_dftb/main.F90]: Discussion regarding removing the save...

File: prog/dftb+/lib_extlibs/negf_int.F90 [hunk 11]
```diff
@@ -0,0 +1,2166 @@
...
+
+  !------------------------------------------------------------------------------
+  ! INTERFACE subroutine to call ldos computation
+  !------------------------------------------------------------------------------
+  subroutine negf_ldos(HH,SS,spin,kpoint,wght,ledos)
+    type(z_CSR), intent(in) :: HH, SS
+    integer, intent(in) :: spin      ! spin index
+    integer, intent(in) :: kpoint        ! kp index
+    real(dp), intent(in) :: wght      ! kp weight
+    real(dp), dimension(:,:), pointer :: ledos
```
Comment (bhourahine) [inline prog/dftb+/lib_extlibs/negf_int.F90]: Does this need to be a pointer?

File: prog/dftb+/lib_extlibs/negf_int.F90 [hunk 12]
```diff
@@ -0,0 +1,2166 @@
...
+  !------------------------------------------------------------------------------
+  ! INTERFACE subroutine to call current computation
+  !------------------------------------------------------------------------------
+  subroutine negf_current(HH,SS,spin,kpoint,wght,tunn,ledos,currents)
+    
+    type(z_CSR), intent(in) :: HH, SS
+    integer, intent(in) :: spin      ! spin index
+    integer, intent(in) :: kpoint        ! kp index
+    real(dp), intent(in) :: wght      ! kp weight
+    real(dp), dimension(:,:), pointer :: tunn
```
Comment (bhourahine) [inline prog/dftb+/lib_extlibs/negf_int.F90]: Possible to be allocatable instead?

File: prog/dftb+/lib_extlibs/negf_int.F90 [hunk 13]
```diff
@@ -0,0 +1,2166 @@
...
+    integer, intent(in) :: iNeighbor(0:,:), nNeighbor(:)
+    integer, intent(in) :: iAtomStart(:), iPair(0:,:)
+    integer, intent(in) :: img2CentCell(:), iCellVec(:)
+    real(dp), intent(in) :: cellVec(:,:)
+    type(TOrbitals), intent(in) :: orb
+    real(dp), intent(in) :: kPoints(:,:), kWeights(:)
+    real(dp), allocatable, intent(inout) :: ldosTot(:,:)
+    logical, intent(in) :: writeLDOS
+
+    real(dp), pointer    :: ldosMat(:,:)=>null()
```
Comment (bhourahine) [inline prog/dftb+/lib_extlibs/negf_int.F90]: or an allocatable?

File: prog/dftb+/lib_extlibs/negf_int.F90 [hunk 14]
```diff
@@ -0,0 +1,2166 @@
...
+    logical, intent(in) :: writeLDOS
+    logical, intent(in) :: writeTunn
+    ! We need this now for different fermi levels in colinear spin
+    ! Note: the spin polirized does not work with
+    ! built-int potentials (the unpolarized does) in the poisson
+    ! I do not set the fermi because it seems that in libnegf it is
+    ! not really needed
+    real(dp), intent(in) :: mu(:,:)
+    
+    real(dp), pointer    :: tunnMat(:,:)=>null()
```
Comment (bhourahine) [inline prog/dftb+/lib_extlibs/negf_int.F90]: or allocatable?

File: prog/dftb+/lib_extlibs/negf_int.F90 [hunk 15]
```diff
@@ -0,0 +1,2166 @@
...
+    end if
+   
+  end subroutine calc_current
+
+  !----------------------------------------------------------------------------
+  !   utility to allocate and sum partial results
+  !----------------------------------------------------------------------------
+  subroutine add_partial_results(mpicomm, pMat, pTot, pSKRes, iK, nK)
+    type(mpifx_comm), intent(in) :: mpicomm
+    real(dp), pointer :: pMat(:,:)
```
Comment (bhourahine) [inline prog/dftb+/lib_extlibs/negf_int.F90]: allocatable?

File: prog/dftb+/lib_extlibs/negf_int.F90 [hunk 16]
```diff
@@ -0,0 +1,2166 @@
...
+
+    !Save H_dftb_orth.mtr to file
+    !open(12,file='H_dftb_orth.mtr',action="write")
+    !do i=1,N
+    !   write(12,*) H(i,1:N)*Hartree__eV
+    !end do
+    !close(12)
+
+    !Save S_dftb_orth.mtr to file
+    !open(12,file='S_dftb_or
```
Comment (bhourahine) [inline prog/dftb+/lib_extlibs/negf_int.F90]: allocatable?

File: external/xmlf90/m_buffer.f90 [hunk 17]
```diff
@@ -35,7 +35,8 @@ module m_buffer
 interface add_to_buffer
       module procedure add_str_to_buffer
 end interface
-private :: add_char_to_buffer, add_str_to_buffer
+private :: add_str_to_buffer
+!private :: add_char_to_buffer
```
Comment (bhourahine) [inline external/xmlf90/m_buffer.f90]: What was the reason for the removal of the routine?

File: prog/dftb+/lib_dftb/scc.F90 [hunk 18]
```diff
@@ -1115,13 +1171,17 @@ subroutine buildShiftPerAtom_(this, env)
...
+   
+    integer :: ll
 
     this%shiftPerAtom(:) = 0.0_dp
 
   #:if WITH_SCALAPACK
     if (env%blacs%atomGrid%iproc /= -1) then
-      deltaQAtom2D(1:1, 1:size(this%deltaQAtom)) => this%deltaQAtom
-      shiftPerAtom2D(1:1, 1:size(this%shiftPerAtom)) => this%shiftPerAtom
+      ll = size(this%deltaQAtom)
```
Comment (bhourahine) [inline prog/dftb+/lib_dftb/scc.F90]: Why the intermediate variable?

File: prog/dftb+/lib_dftb/scc.F90 [hunk 19]
```diff
@@ -858,11 +882,43 @@ subroutine getShiftPerL(this, shift)
...
 
   end subroutine getShiftPerL
 
+  !> set the shifts from outside (e.g. Poisson solver)
+  subroutine setShiftPerAtom(this, shift)
+
+    !> Instance
+    class(TScc), intent(inout) :: this
+
+    !> Contains the input shifts: shift(atom, spin).
```
Comment (bhourahine) [inline prog/dftb+/lib_dftb/scc.F90]: Comment apparently incorrect, no spin index

File: prog/dftb+/lib_dftb/scc.F90 [hunk 20]
```diff
@@ -858,11 +882,43 @@ subroutine getShiftPerL(this, shift)
     @:ASSERT(this%tInitialised)
     @:ASSERT(size(shift,dim=1) == size(this%shiftPerL,dim=1))
     @:ASSERT(size(shift,dim=2) == size(this%shiftPerL,dim=2))
-    shift = 0.0_dp
+    
     shift = this%shiftPerL
 
   end subroutine getShiftPerL
 
+  !> set the shifts from outside (e.g. Poisson solver)
```
Comment (bhourahine) [inline prog/dftb+/lib_dftb/scc.F90]: Sets electrostatic shift?

File: prog/dftb+/lib_type/typegeometry.F90 [hunk 21]
```diff
@@ -8,10 +8,12 @@
 !> data type and associated routines for specifying atomic geometry and boundary conditions
 module typegeometry
   use accuracy
+  use lapackroutines
   implicit none
   private
 
   public :: TGeometry, normalize
+  public :: reduce, setLattice
```
Comment (bhourahine) [inline prog/dftb+/lib_type/typegeometry.F90]: Can reduce have a more informative name?

File: prog/dftb+/lib_type/typegeometry.F90 [hunk 22]
```diff
@@ -95,4 +105,62 @@ subroutine Geometry_normalize(sf)
 
   end subroutine Geometry_normalize
 
+  !> Reduce the geometry on a subset.
```
Comment (bhourahine) [inline prog/dftb+/lib_type/typegeometry.F90]: Meaning?

File: prog/dftb+/prg_dftb/main.F90 [hunk 23]
```diff
@@ -4394,10 +4731,12 @@ subroutine getNextCoordinateOptStep(pGeoCoordOpt, EMermin, derivss, indMovedAtom
     real(dp) :: derivssMoved(3 * size(indMovedAtom))
     real(dp), target :: newCoordsMoved(3 * size(indMovedAtom))
     real(dp), pointer :: pNewCoordsMoved(:,:)
+    integer :: ll
 
     derivssMoved(:) = reshape(derivss(:, indMovedAtom), [3 * size(indMovedAtom)])
     call next(pGeoCoordOpt, EMermin, derivssMoved, newCoordsMoved, tCoordEnd)
-    pNewCoordsMoved(1:3, 1:size(indMovedAtom)) => newCoordsMoved(1 : 3 * size(indMovedAtom))
+    ll = size(indMovedAtom)
```
Comment (bhourahine) [inline prog/dftb+/prg_dftb/main.F90]: Why the intermediate variable?

File: sys/make.x86_64-linux-intel [hunk 24]
```diff
@@ -50,7 +50,7 @@ FXX = ifort
 CC = icc
 
 # Compile options 
-FXXOPT = -O2 -qopenmp -ip -standard-semantics -heap-arrays 10
```
Comment (bhourahine) [inline sys/make.x86_64-linux-intel]: Why?

File: test/prog/dftb+/transport/CNT/dftb_pin.hsd [hunk 25]
```diff
@@ -0,0 +1,278 @@
...
+18    1      5.010000000000      0.527000000000     -7.380000000000
+19    1      5.010000000000      4.511000000000     -7.380000000000
+20    1      5.583000000000      3.223000000000     -7.380000000000
+}
+Transport = {
+  Device = {
+    AtomRange = {
+1 80
+    }
+    FirstLayerAtoms = {
```
Comment (bhourahine) [inline test/prog/dftb+/transport/CNT/dftb_pin.hsd]: Should this not be 1 to 40 ?

File: prog/dftb+/lib_type/typegeometry.F90 (hunk 21 above)
Comment (aradi) [reply prog/dftb+/lib_type/typegeometry.F90]: It reduces the full geometry to a subset (e.g. contact). :wink:  If you have any better suggestion, feel free to rename.

File: prog/dftb+/prg_dftb/parser.F90 [hunk 26]
```diff
@@ -3374,6 +3520,1316 @@ subroutine readCustomisedHubbards(node, geo, orb, tShellResolvedScc, hubbU)
...
+    end if
+
+    ! Determine to which axes it is parallel.
+    mask = (abs(abs(contactVec)  - sqrt(sum(contactVec**2))) < 1.0e-8_dp)
+    if (count(mask) /= 1) then
+      call warning("Contact vector " // i2c(id) // " not parallel to any&
+          & of the coordinate axis.")
+      contactDir = 0
+    else
+      ! Workaround for bug in Intel compiler (can not use index function)
```
Comment (bhourahine) [inline prog/dftb+/prg_dftb/parser.F90]: Which compiler version - is this still a problem?

**Pull request merged**